llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -8846,9 +8846,14 @@ static bool findGISelOptimalMemOpLowering(std::vector<LLT> &MemOps,
   if (Op.isMemcpyWithFixedDstAlign() && Op.getSrcAlign() < Op.getDstAlign())
     return false;
 
-  LLT Ty = TLI.getOptimalMemOpLLT(Op, FuncAttributes);
-
-  if (Ty == LLT()) {
+  bool WantIntScalar = TLI.useSoftFloat() ||
+                       FuncAttributes.hasFnAttr(Attribute::NoImplicitFloat);
```

Comment:
I'd prefer to avoid eagerly querying these edge cases. Also this seems like another case where we have overlapping and poorly defined mechanisms. For example useSoftFloat is a virtual function without any arguments, which is kind of useless. Why isn't it a simple function field, and isn't it a function of the use-soft-float attribute, so it can't just be part of the subtarget's TargetLowering? 

At least could just merge these into a helper for implementations of getOptimalMemOpLLT to use?

Reply:
@arsenm ping for clarification on this point:

> Do you mean that you would prefer if I just fixed each `getOptimalMemOpLLT()`/`getOptimalMemOpType()` implementation to check for these attributes when they want to potentially return a float/vector type?

Reply:
Yes. Can put the query into a helper somewhere for the implementations to use?